stdlib/private/TensorFlowUnittest/TensorFlowUnittestRemote.swift

```diff
@@ -0,0 +1,109 @@
+#if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
+import Darwin
+#else
+import Glibc
+#endif
+import CTensorFlow
+import CTensorFlowUnittest
+import TensorFlow
+import StdlibUnittest
+
+typealias CTFServer = OpaquePointer
+
+@usableFromInline
+func debugLog(_ message: @autoclosure () -> String,
+              file: StaticString = #file,
+              line: UInt = #line) {
+  if _RuntimeConfig.printsDebugLog {
+    print("[\(file):\(line)] \(message())")
+    // This helps dump more log before a crash.
+    fflush(stdout)
+  }
+}
+
+
+private class TestCluster {
+  var num_tasks:Int = 0
+  var servers: [CTFServer] = []
+  public var serverDef:String = ""
+
+  public init(num_tasks: Int) {
+    self.num_tasks = num_tasks
+    self.serverDef = startServers()
+  }
+
+  deinit {
+    // TODO: There is no clean way to shutdown GrpcServers yet.
+  }
+
+  private func getServerDef(tasks: String, task_index: Int) -> String {
+    return  """
+      cluster {
+        job {
+          name: "localhost"
+          \(tasks)
+        }
+      }
+      job_name: "localhost"
+      task_index: \(task_index)
+      protocol: "grpc"
+      """
+  }
+
+  private func getTasks() -> String {
+    var tasks = ""
+    for i in 0..<num_tasks {
+      let port = TF_PickUnusedPortOrDie()
+      let task = """
+          tasks {
+            key: \(i)
+            value: "localhost:\(port)"
+          }
+      """
+      debugLog("Picking port \(port) for \(i)...");
+      tasks += task
+    }
+    return tasks;
+  }
+
+  private func startServers() -> String {
+    let tasks = getTasks()
+    let status = TF_NewStatus()
+    // Start everything except the 0-th task, which will
+    // be started when we set TFE_ContextSetServerDef.
+    for i in 1..<num_tasks {
+      debugLog("Starting task \(i)...");
+      let serverDefText = getServerDef(tasks: tasks, task_index: i)
+      let serverDef = TFE_GetServerDef(serverDefText, status)
+      checkOk(status)
+      let server = TF_NewServer(
+          serverDef!.pointee.data, serverDef!.pointee.length, status)
+      checkOk(status)
+      TF_ServerStart(server, status)
+      checkOk(status)
+      servers.append(server!)
+      let target = TF_ServerTarget(server!)
+      debugLog("Started task \(i) at \(String(cString: target!))")
+    }
+    TF_DeleteStatus(status)
+    // Return the server def for task 0 for use in compiler runtime.
+    return getServerDef(tasks: tasks, task_index: 0)
+  }
+}
+
+public func runAllTestsWithRemoteSession(num_tasks: Int = 2) {
```

Comment:
should we switch over some unit tests to use this remote mode?

Reply:
Yes, I think we should test both paths in all tests? Will send out a separate PR for that.

Reply:
If the tests are very cheap, then yes. But in our case I believe it'd be better to pick a few test suites (or even just a single test) and run remote. This'll keep test execution lean without losing coverage.

Reply:
Enabled remote tests in TensorFlowRuntime/tensor.swift. PTAL.